test/BlockStrings.hs

```diff
@@ -0,0 +1,96 @@
+
+module BlockStrings where
+
+import           Data.Attoparsec.Text
+import qualified Data.Text                      as T
+import qualified Data.Text.Prettyprint.Doc      as PP
+import           Data.Void                      (Void)
+import           Hedgehog
+import           Language.GraphQL.Draft.Parser
+import qualified Language.GraphQL.Draft.Printer as Printer
+import           Language.GraphQL.Draft.Syntax
+import qualified Prettyprinter.Render.Text      as PP
+
+blockTest :: IO Bool
+blockTest = do
+  checkParallel $ Group "Test.parser.block-string.unit"
+    [ ("parses the specExample", blockParsesTo "\n    Hello,\n      World!\n\n    Yours,\n      GraphQL.\n  " "Hello,\n  World!\n\nYours,\n  GraphQL.")
+    , ("do not remove WS from the end of lines", blockParsesTo "\nFoo \nbar  " "Foo \nbar  ")
+    , ("tabs are WS as well", blockParsesTo "\n\t\tFoo\n\t\tbar\n\t\t\tqux" "Foo\nbar\n\tqux")
+    , ("tabs work with spaces", blockParsesTo "\n\t Foo\n \tbar\n\t\t qux" "Foo\nbar\n qux")
+    , ("parses newline", blockParsesTo "\n" "")
+    , ("parses very simples not-empty block", blockParsesTo "x" "x")
+    , ("common indentation is removed", blockParsesTo "\n  a \n   b \n  c " "a \n b \nc ")
+    , ("zero common indentation is possible", blockParsesTo "\na \n b \nc " "a \n b \nc ")
+    , ("no whitespace is removed from the first line", blockParsesTo "  abc " "  abc ")
+    , ("ignores escaping", blockParsesTo "  \\  " "  \\  ") -- this is a single \
+    , ("\n in first characters is parsed", blockParsesTo "\n hey  " "hey  ")
+    , ("", blockParsesTo "\nx\n" "x")
+    , ("empty single line", blockParsesTo "" "")
+    , ("empty two lines", blockParsesTo "\n" "")
+    , ("empty three lines", blockParsesTo "\n\n" "")
+    , ("empty X lines", blockParsesTo "\n\n\n\n\n\n" "")
+    , ("preserves escaped newlines", blockParsesTo "\nhello\\nworld\n" "hello\\nworld")
+    , ("", blockParsesTo "\n\"\n" "\"")
+    , ("escaped triple-quotes are ignored as block terminator", blockParsesTo "\n   \\\"\"\"hey\n   friends\n" "\\\"\"\"hey\nfriends")
+    , ("fails for normal string", blockParseFail (\t -> "\"" <> t <> "\"") "hey")
+    , ("fails for block string that is not closed", blockParseFail ("\"\"\"" <>) "hey")
+    , ("fails for block string that is not closed when there are escaped triple-quotes",
+          blockParseFail (\t -> "\"\"\"" <> t <> "\\\"\"\"" <> t) "hey")
+    , ("does not ignore escaping when its part of a escaped tripel-quotes",
+          blockParseFail (\t -> "\"\"\"" <> t <> "\"\"\"") "\\") -- this is a single \, but it touches the """ at the end
+
+    -- Printing
+    , ("a line with indentation prints as blockstring", printsAsBlockString (VString "foo\n    indented!"))
+    , ("when starts with indentation prints a regular string", printsAsString (VString "    indented!"))
+    , ("when has escaped newlines prints as regular string", printsAsString (VString "foo\\n    indented!"))
```

Comment:
What's wrong with the following?
```
"""
foo\n    indented!
"""
```

Reply:
To be honest, these `printsAsBlockString` and `printsAsString` test cases do not make me very convinced that the code is free of bugs.

Reply:
This was discussed in the call. We decided to err on the safe side when printing strings and choose to use a regular string if we detected any escaped new lines.

Reply:
To put it differently: a test suite should test when _any_ implementation could be considered (in)correct, rather than verifying what the expected behavior of a known implementation is.

Reply:
Got it, that makes sense to me if you want to use tests as documentation about the standard. 

What type of tests do you suggest we have to show that we are indeed printing things as block texts?

Reply:
I think it suffices the following two categories of tests suffice:
- roundtrip tests, namely that `parse . print = id` (the other way around is not important for us) (this is covered by `genBlockText`), and
- parsing tests, namely that a specific block string is interpreted as a specific Haskell text value (this is covered by `blockTest`).

I don't think it is crucial to test that the printer works correctly in isolation - so just remove the last 4 lines in `blockTest`.